### DISCUSSION TITLE
Crash - "Start may not be called on a task that has completed" - at "OnProcessMessageReceived"

### DISCUSSION THREAD

**Discussion by eadan-y** [cefsharp/CefSharp#3638] [General]
Hey
We see a crash from our end-users.
Wasn't able to reproduce this locally yet so no information other than the stack trace
CefSharp version - 90.6.50

System.InvalidOperationException: Start may not be called on a task that has completed.\r\n   at System.Threading.Tasks.Task.Start(TaskScheduler scheduler)\r\n   at CefSharp.Internals.ConcurrentMethodRunnerQueue.Enqueue(MethodInvocation methodInvocation)\r\n   at CefSharp.Internals.ClientAdapter.OnProcessMessageReceived(ClientAdapter* , scoped_refptr<CefBrowser>* browser, scoped_refptr<CefFrame>* frame, cef_process_id_t source_process, scoped_refptr<CefProcessMessage>* message)

**Comment by amaitland** [comment]
Nothing immediately stands out as to the cause. Do you have any clues as to what was happening when the exception occurred? Was Dispose being called? Has the exception occurred more than once?

My initial thoughts are we should switch to using `Task.Run` as we are simply queueing on the `ThreadPool`, no reason to use `Task.Start` (or at least I don't remember there being a reason).

I'll see if `Google` yields any possible causes based on the error message.

**Reply by eadan-y** [reply]
It is very likely that this was during a Dispose of the browser, similar to ([https://github.com/cefsharp/CefSharp/discussions/3573]). We have a communication between the C# client and the browser through JS Repository object with async functions, and while they are being called, the browser window can be closed. 

We've seen it more than once, but it is quite rare.

**Reply by eadan-y** [reply]
Is it possible that another thread has canceled the task (through it's cancellation token) right after creating the task but before running it?(https://github.com/cefsharp/CefSharp/blob/4608924d5a0c676857720917550674b88bd7b701/CefSharp/Internals/ConcurrentMethodRunnerQueue.cs#L116)

**Reply by amaitland** [reply]
> Is it possible that another thread has canceled the task (through it's cancellation token) right after creating the task but before running it?

The `Dispose` method is called on a different thread so it's possible the `Task` is cancelled before it's started. It should be trivial to confirm that sequence of events results in the exception you are seeing.

If that is the case then switching to `Task.Run` should resolve the issue.

**Reply by amaitland** [reply]
Removing the `cancellationTokenSource.IsCancellationRequested` check from [ConcurrentMethodRunnerQueue.Enqueue](https://github.com/cefsharp/CefSharp/blob/cefsharp/91/CefSharp/Internals/ConcurrentMethodRunnerQueue.cs#L37) and running the below `xUnit` test I get an `InvalidOperationException`.

```c#
[Fact]
public void DisposeConcurrentMethodRunnerQueueThenQueueTaskDemo()
{
	var methodInvocation = new MethodInvocation(1, 1, 1, "Testing", 1);
	methodInvocation.Parameters.Add("Echo Me!");

	var objectRepository = new JavascriptObjectRepository
	{
		NameConverter = null
	};

	var methodRunnerQueue = new ConcurrentMethodRunnerQueue(objectRepository);

	methodRunnerQueue.Dispose();

	Assert.Throws<InvalidOperationException>(() => methodRunnerQueue.Enqueue(methodInvocation));
}
```

I created a couple of synthetic tests as well to compare the behaviour of using `Task.Start` vs `Task.Run`. Will switch to `Task.Run` as the result is a `Task` with `TaskStatus.Canceled`, no exception is thrown.

```c#
[Fact]
public void SimulateStartOnTaskAlreadyCompleted()
{
	var cts = new CancellationTokenSource();

	//Create a new Task
	var task = new Task(async () =>
	{
		await Task.Delay(100);

	}, cts.Token);

	//Cancel before started
	cts.Cancel();

	Assert.Throws<InvalidOperationException>(() => task.Start(TaskScheduler.Default));
}

[Fact]
public void SimulateTaskRunStartOnTaskAlreadyCompleted()
{
	var cts = new CancellationTokenSource();

	//Cancel before started
	cts.Cancel();

	var task = Task.Run(async () =>
	{
		await Task.Delay(100);
	}, cts.Token);

	Assert.NotNull(task);
	Assert.Equal(TaskStatus.Canceled, task.Status);
}
```

**Reply by amaitland** [reply]
I've created #3639 (so the issue can be assigned to a milestone) and committed my proposed fix in https://github.com/cefsharp/CefSharp/commit/4ab6d2ad25cb0dc19ced4dc0dad040a863c1a616

**Reply by amaitland** [reply]
@eadan-y Let me know if you are happy and I can release a new version. Can confirm if you are using version 90 or 91?

**Reply by eadan-y** [reply]
@amaitland 
We currently use 90 in production and 91 in the next version. Fixing it for 91 should be enough. Thanks for the quick response!

**Reply by amaitland** [reply]
Version `91.1.211` is now on `Nuget.org`
